torch/distributed/_checkpoint/state_dict_saver.py

```diff
@@ -0,0 +1,189 @@
+    ShardedTensor,
+    ShardedTensorMetadata,
+    TensorProperties,
+    ShardMetadata,
+)
+
+from .metadata import (
+    Metadata,
+    BytesWriteRequest,
+    ExtendedTensorMetadata,
+    StorageMetadata,
+    TensorWriteRequest,
+)
+from .resharding import prepare_sharded_tensor_write
+from .storage_writer import StorageWriter
+
+# -------------- private functions --------------
+def _compute_tensor_md(fqn: str, tensor: Tensor) -> ExtendedTensorMetadata:
+    # --- Step 3, populate the metadata ---
+    #
+    # Since torch.Tensor does not have a standard set of metadata we can operate on
+    # We wrap troch.Tensor's metadata with ShardMetadata
+    # This is frankly a bad idea, I will need to change this
+    tensor = tensor.detach()
+    tensor_size = list(tensor.size())
+    storage_size = tensor.nelement() * tensor.element_size()
+    shard_metadata = ShardMetadata(
+        shard_offsets=[0] * len(tensor_size),
+        shard_sizes=tensor_size,
+        # Not sure how to deal with placement for regular tensor yet.
+        # Since they are only keep the copy on rank0, let's hard code it for now.
+        placement=f"rank:0/{str(tensor.device)}",
+    )
+
+    stm = ShardedTensorMetadata(
+        shards_metadata=[shard_metadata],
+        size=tensor.size(),
+        tensor_properties=TensorProperties(
+            dtype=tensor.dtype,
+            layout=tensor.layout,
+            requires_grad=tensor.requires_grad,
+            memory_format=torch.contiguous_format,
+            pin_memory=tensor.is_pinned(),
+        ),
+    )
+
+    return ExtendedTensorMetadata(
+        tensor_metadata=stm,
+        storage_metadata=[
+            StorageMetadata(
+                shard_metadata=shard_metadata,
+                storage_key=fqn,
+                length=storage_size,
+                offset=0,
+            )
+        ],
+    )
+
+
+def _prepare(
+    state_dict: Dict[str, Any], always_add_tensors: bool = False
+) -> Tuple[Metadata, Dict[str, int], List[BytesWriteRequest], List[TensorWriteRequest]]:
+    """
+    Uses the state_dict to build three things.
+
+    metadata: Metadata
+        The metatdata discribing the tensor / sharded tensor.
+        And it is storage meta data. See "../metadata.py" for detail
+
+    size_for_storage_keys: Dict[str, int]
+        Key is the storage key name, value is its size
+        It can used to pre allocate the storage for parallel and non sequential writes.
+
+    tensor_write_requests: List[TensorWriteRequest]
+        List of tensor write requests that should br perfromed by the writer.
+
+    bytes_write_requests: List[BytesWriteRequest]
+        List of byte write requests that should br perfromed by the writer.
+
+    Subclasses can optionally overwrite the implementation here,
+    if the default does not meet its requirement.
+    Args:
+        state_dict: The state_dict to operate on
+        always_add_tensors: Include non-sharded tensors even if rank != 0
```

Comment:
It's for testing. This allows to write tests that don't require to be on rank0.

Reply:
I don't think we should add flags like this in the code just for testing purposes. We can test this accordingly in unit tests by doing things like `if self.rank == 0: validate_something()`. I feel the latter is more robust since otherwise your unit tests are testing a codepath that is never actually triggered in practice.